RecoLocalTracker/Phase2TrackerRecHits/plugins/VectorHitBuilderEDProducer.cc

```diff
@@ -0,0 +1,134 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+class VectorHitBuilderEDProducer : public edm::stream::EDProducer<>
+{
+
+ public:
+
+  explicit VectorHitBuilderEDProducer(const edm::ParameterSet&);
+  virtual ~VectorHitBuilderEDProducer();
+  virtual void produce(edm::Event&, const edm::EventSetup&) override;
+  void setupAlgorithm(edm::ParameterSet const& conf);
+  void run(edm::Handle< edmNew::DetSetVector<Phase2TrackerCluster1D> > clusters,
+           edmNew::DetSetVector<Phase2TrackerCluster1D>& clustersAcc, edmNew::DetSetVector<Phase2TrackerCluster1D>& clustersRej,
+           VectorHitCollectionNew& outputAcc, VectorHitCollectionNew& outputRej);
+//  VectorHitBuilderAlgorithm * algo() const { return builderAlgo; };
+
+ private:
+  VectorHitBuilderAlgorithm* builderAlgo;
+  std::string offlinestubsTag;
+  unsigned int maxOfflinestubs;
+  std::string algoTag;
+  edm::EDGetTokenT< edmNew::DetSetVector<Phase2TrackerCluster1D> > clusterProducer;
+  bool readytobuild;
+
+};
+
+VectorHitBuilderEDProducer::VectorHitBuilderEDProducer(edm::ParameterSet const& conf)
+  : offlinestubsTag( conf.getParameter<std::string>( "offlinestubs" ) ),
+    maxOfflinestubs(conf.getParameter<int>( "maxVectorHits" )),
+    algoTag(conf.getParameter<std::string>( "Algorithm" )),
+    readytobuild(false)
+{
+
+  clusterProducer = consumes< edmNew::DetSetVector<Phase2TrackerCluster1D> >(edm::InputTag(conf.getParameter<std::string>("Clusters")));
+
+  produces< edmNew::DetSetVector< Phase2TrackerCluster1D > >( "ClustersAccepted" );
+  produces< edmNew::DetSetVector< Phase2TrackerCluster1D > >( "ClustersRejected" );
+  produces< VectorHitCollectionNew >( offlinestubsTag + "Accepted" );
+  produces< VectorHitCollectionNew >( offlinestubsTag + "Rejected" );
+  setupAlgorithm(conf);
+}
+
+VectorHitBuilderEDProducer::~VectorHitBuilderEDProducer() {
+  delete builderAlgo;
+}
+
+void VectorHitBuilderEDProducer::produce(edm::Event& event, const edm::EventSetup& es)
+{
+  LogDebug("VectorHitBuilder") << "VectorHitBuilderEDProducer::produce() begin";
+
+  // get input clusters data
+  edm::Handle< edmNew::DetSetVector<Phase2TrackerCluster1D> >  clustersHandle;
+  event.getByToken( clusterProducer, clustersHandle);
+
+  // create the final output collection
+  std::unique_ptr< edmNew::DetSetVector< Phase2TrackerCluster1D > > outputClustersAccepted( new edmNew::DetSetVector< Phase2TrackerCluster1D > );
+  std::unique_ptr< edmNew::DetSetVector< Phase2TrackerCluster1D > > outputClustersRejected( new edmNew::DetSetVector< Phase2TrackerCluster1D > );
+  std::unique_ptr< VectorHitCollectionNew > outputVHAccepted( new VectorHitCollectionNew() );
+  std::unique_ptr< VectorHitCollectionNew > outputVHRejected( new VectorHitCollectionNew() );
+
+  if(readytobuild)  builderAlgo->initialize(es);
+  else edm::LogError("VectorHitBuilder") << "Impossible initialization of builder!!";
+
+  // check on the input clusters
+  builderAlgo->printClusters(*clustersHandle);
```

Comment:
If the only effect of `printClusters` are debug printouts, enclose the call inside `#ifdef EDM_ML_DEBUG` so that it does not get run unless when debugging.